### PR TITLE
Update Elastica for Elasticsearch 7

### PR DESCRIPTION
**Pull request by p365labs**
This PR is to update the compatibility for Elastic to Elasticsearch 7.

The document followed to update Elastica is the one on [ES7 Breaking Changes](https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html).

### CODE REVIEW

File: env/elasticsearch/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #https://www.docker.elastic.co/
-FROM docker.elastic.co/elasticsearch/elasticsearch:6.5.2
+FROM docker.elastic.co/elasticsearch/elasticsearch:7.0.0-alpha1
```
Comment (ruflin) [inline env/elasticsearch/Dockerfile]: Lets use `docker.elastic.co/elasticsearch/elasticsearch:7.0.0-alpha1-SNAPSHOT` during the development for this. It will make the CI more flaky here as it could be that new breaking changes just came in but lots of changes have happened since `7.0.0-alpha1`.

File: test/Elastica/Aggregation/SignificantTermsTest.php [hunk 2]
```diff
@@ -15,17 +15,17 @@ protected function _getIndexForTest()
         $colors = ['blue', 'blue', 'red', 'red', 'green', 'yellow', 'white', 'cyan', 'magenta'];
         $temperatures = [1500, 1500, 1500, 1500, 2500, 3500, 4500, 5500, 6500, 7500, 7500, 8500, 9500];
 
-        $mapping = new Mapping($index->getType('test'), [
+        $mapping = new Mapping($index->getType('_doc'), [
             'color' => ['type' => 'keyword'],
             'temperature' => ['type' => 'keyword'],
         ]);
-        $index->getType('test')->setMapping($mapping);
+        $index->getType('_doc')->setMapping($mapping);
```
Comment (ruflin) [inline test/Elastica/Aggregation/SignificantTermsTest.php]: I'm wondering if we should just do a separate PR to do all the `_doc` conversion. This should already work against 6.x and would make this PR much smaller.

**Comment by ruflin** [comment]
@p365labs More general question: Should we start to create a 6.x branch and all new development goes into master and if needed in 6.x needs to be backported?

File: env/elasticsearch/Dockerfile (hunk 1 above)
Comment (p365labs) [reply env/elasticsearch/Dockerfile]: @ruflin I'm facing this issue with the snapshot, and I didn't find documentation about it :

```
bin/elasticsearch-plugin install --batch ingest-attachment
ERROR: attempted to install release build of official plugin on snapshot build of Elasticsearch
```

Do you have an idea about it ?

Comment (ruflin) [reply env/elasticsearch/Dockerfile]: Perhaps the line here helps (different plugin, but same mechanism should work): https://github.com/elastic/beats/blob/6.4/testing/environments/docker/elasticsearch/Dockerfile#L39

Comment (p365labs) [reply env/elasticsearch/Dockerfile]: yes it's useful but I'm managing to find the right URL :)

this is not working
```
https://snapshots.elastic.co/downloads/elasticsearch-plugins/ingest-attachment/ingest-attachment-7.0.0-alpha1-SNAPSHOT.zip```

any hint on this ?

Comment (p365labs) [reply env/elasticsearch/Dockerfile]: @ruflin :) finally I've been able to build the URL 👍 

```https://snapshots.elastic.co/downloads/elasticsearch-plugins/ingest-attachment/ingest-attachment-7.0.0-SNAPSHOT.zip```

**Comment by p365labs** [comment]
I need to : 

- [ ] remove the _doc commit
- [ ] fix the way ingest-attachment plugin, for SNAPSHOT image, is downloaded

File: test/Elastica/Aggregation/SignificantTermsTest.php (hunk 2 above)
Comment (p365labs) [reply test/Elastica/Aggregation/SignificantTermsTest.php]: done #1574 :)

**Comment by ruflin** [comment]
Other PR is merged, we need to rebase this one now.

Not in this PR but we should also deprecate getType() probably?

**Comment by ruflin** [comment]
@p365labs I see you got this one green. I wonder how soonish we should merge this one. We need for sure have the 6.x branch first.

**Comment by p365labs** [comment]
@ruflin yep, we need to merge all the branches about type hinting first, then create a branch 6.x from master, and then we can merge ES7 branch on master and using master for the development of the new Major (7.x). what do u think ?

**Comment by ruflin** [comment]
SGTM

**Comment by p365labs** [comment]
@ruflin tests are failing because of this change [between 6.7 and 7.x](https://www.elastic.co/guide/en/elasticsearch/reference/master/removal-of-types.html#_schedule_for_removal_of_mapping_types):
- in **6.7** _include_type_name_ is true by default
- in **7.x** _include_type_name_ is false by default

The problem with this is that [elasticsearch-php](https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/Endpoints/Index.php#L82) should add to whitelist parameters this include_type_name which at the moment don't...

What about updating tests removing types (as it has been deprecated and then removed in 8.x)? 
This will allow us to pass tests and in general, avoid the use of _type_

What di u think ?

**Comment by ruflin** [comment]
+1 on removing tests which need / require tests. For elasticsearch-php, should we open an issue?

**Comment by p365labs** [comment]
@ruflin ok,
I've already a pr, ready for it...but lack of time ;)
I think I'll do it this evening.

**Comment by ruflin** [comment]
@p365labs Now that 7.0 is out we should probably update to the most recent release version :-) Do you have any guess how much effort this migration is?

**Comment by p365labs** [comment]
@ruflin yes make a lot of sense, I'm working on this during the last couple of days and I'm counting to deliver something that works for the next weekend. Make sense?

Sorry for being so late, but this is a really busy period :)

**Comment by keywinf** [comment]
Hi, is search_after going to be supported ? https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-search-after.html. It seems unusable with 6.x versions

**Comment by dinamic** [comment]
@p365labs let us know if you need a helping hand. Maybe have some checkboxes with TODOs to indicate what else needs to be worked on?

**Comment by p365labs** [comment]
@ruflin I think this PR is wainting too much in order to be merged. I mean right now tests are green and the only things we are waiting is elasticsearch-php to release a version for 7.x. We are using ```        "elasticsearch/elasticsearch": "dev-master"```

For me this could not be a problem. My suggestion is to merge this branch and when elasticsearch-ph will do a release we will update composer.

This will allow people using Elastica testing it with ES 7.x and as suggested by @dinamic we could create an issue with task in order to implement all the new features of ES7 and allow people also to help us doing it.

What do u think?

File: env/elasticsearch/Dockerfile [hunk 3]
```diff
@@ -1,8 +1,9 @@
 #https://www.docker.elastic.co/
-FROM docker.elastic.co/elasticsearch/elasticsearch:6.6.1
+FROM docker.elastic.co/elasticsearch/elasticsearch:7.0.0
 MAINTAINER Nicolas Ruflin <spam@ruflin.com>
 
 RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install --batch ingest-attachment
+#RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install --batch https://snapshots.elastic.co/downloads/elasticsearch-plugins/ingest-attachment/ingest-attachment-7.0.0-SNAPSHOT.zip
```
Comment (ruflin) [inline env/elasticsearch/Dockerfile]: Could you remove this line?

File: env/elasticsearch/Dockerfile [hunk 4]
```diff
@@ -1,8 +1,9 @@
 #https://www.docker.elastic.co/
-FROM docker.elastic.co/elasticsearch/elasticsearch:6.6.1
+FROM docker.elastic.co/elasticsearch/elasticsearch:7.0.0
```
Comment (ruflin) [inline env/elasticsearch/Dockerfile]: Should we directly go to 7.1.0?

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,11 +6,18 @@ All notable changes to this project will be documented in this file based on the
...
+* [Camel Case](https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html#_camel_case_and_underscore_parameters_deprecated_in_6_x_have_been_removed) and underscore parameters deprecated in 6.x have been removed
+* The parameter [fields](https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html#_the_parameter_literal_fields_literal_deprecated_in_6_x_has_been_removed_from_bulk_request) deprecated in 6.x has been removed from Bulk requestedit and Update request.
+* The [_parent](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-parent-field.html) field has been removed in favour of the join field.
+* hits.total is now an object in the search responseedit [hits.total](https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html#_literal_hits_total_literal_is_now_an_object_in_the_search_response)
```
Comment (ruflin) [inline CHANGELOG.md]: responseedit?

**Comment by p365labs** [comment]
@ruflin done

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,11 +6,18 @@ All notable changes to this project will be documented in this file based on the
 ### Backward Compatibility Breaks
 * \Elastica\AbstractUpdateAction::getOptions( $fields ) no longer supports the $underscore parameter, option names must match what elasticsearch expects.
 * Removed no longer supported \Elastica\Query\QueryString::setAutoGeneratePhraseQueries( $bool ) [#1622](https://github.com/ruflin/Elastica/pull/1622)
+* Replaced [params._agg](https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html#_replaced_literal_params__agg_literal_with_literal_state_literal_context_variable_in_scripted_metric_aggregations) with state context variable in scripted metric aggregations
```
[the remark this answers is not in the document]
Comment (p365labs) [reply CHANGELOG.md]: removed

File: CHANGELOG.md (hunk 5 above)
Comment (p365labs) [reply CHANGELOG.md]: yep was a typo... it's only **response**

File: env/elasticsearch/Dockerfile (hunk 4 above)
Comment (p365labs) [reply env/elasticsearch/Dockerfile]: yep u're right! switched to 7.1.0

File: env/elasticsearch/Dockerfile (hunk 3 above)
Comment (p365labs) [reply env/elasticsearch/Dockerfile]: done.

**Comment by ruflin** [comment]
@p365labs Merged. I also update the description of the PR and took it as commit message. From now on there is only moving forward to 7.0 :-)

**Comment by ruflin** [comment]
For awareness, elasticsearch-php 7.0 was just released: https://github.com/elastic/elasticsearch-php/releases/tag/v7.0.0 Thanks @ezimuel
